platforms/android/tangram/src/main/java/com/mapzen/tangram/MapView.java

```diff
@@ -18,45 +19,130 @@
+     * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
+     */
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback) {
+        return getMapAsync(readyCallback, null, null);
+    }
+
+    /**
+     * Construct a {@code MapController} in an async thread; may only be called from the UI thread
      * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
-     * @param listener The listener to receive to receive scene load events;
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
+     * @param sceneLoadListener The listener to receive to receive scene load events;
      * the callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener) {
-        return getMap(listener, null);
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback,
+                            @Nullable final MapController.SceneLoadListener sceneLoadListener) {
+        return getMapAsync(readyCallback, sceneLoadListener, null);
     }
 
     /**
-     * Construct a {@code MapController}; may only be called from the UI thread
-     * @param listener The listener to receive to receive scene load events;
+     * Construct a {@code MapController} in an async thread; may only be called from the UI thread
+     * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
+     * @param sceneLoadListener The listener to receive to receive scene load events;
      * @param handler Set the client implemented {@link HttpHandler} for retrieving remote map resources
      *                when null {@link DefaultHttpHandler} is used
-     * the callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener, @Nullable final HttpHandler handler) {
-        if (mapController != null) {
-            return mapController;
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback,
```

Comment:
Also, on the early return path we probably don't want to invoke the `MapReadyCallback` directly in this method.  In my experience it can be confusing or disruptive to have a supposedly asynchronous method invoke its completion handler before the original synchronous call returns. Instead we could allow the async task to run and just skip most of the work when the `MapController` is already initialized.

Reply:
Yeah will update the client should be able to keep a reference of the instantiated `MapController` through the `onMapReady` interface.